spread/testflinger.go

```diff
@@ -0,0 +1,268 @@
+            Url: image,
+        },
+        TestData: &tfTestData{
+            TestCommands: []string{"ssh $DEVICE_IP cat /proc/cpuinfo", "ssh $DEVICE_IP echo ubuntu:ubuntu | sudo chpasswd", "ssh $DEVICE_IP echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/create-user-ubuntu"},
+        },
+        ReserveData: &tfReserveData{
+            SSHKeys: []string{key},
+        },
+    }
+
+    d, err := yaml.Marshal(file)
+    if err != nil {
+        return nil, fmt.Errorf("error: %v", err)
+    }
+
+	tmpfile, err := ioutil.TempFile(".", tfconfigfile)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating temporal file: %v", err)
+	}
+	defer os.Remove(tmpfile.Name())
+
+	if _, err := tmpfile.Write(d); err != nil {
+		return nil, fmt.Errorf("Error writing temporal file: %v", err)
+	}
+	if err := tmpfile.Close(); err != nil {
+		return nil, fmt.Errorf("Error closing temporal file: %v", err)
+	}
+
+	// First step is to get the job_id running the submit command
+    jobId := ""
+
+    out, err := exec.Command("/snap/bin/testflinger-cli", "submit", tmpfile.Name()).Output()
+    if err != nil {
+        return nil, fmt.Errorf("Error running command: %v: ", err)
+    }
+    lines := strings.Split(string(out), "\n")
+
+    for _, line := range lines {
+        if strings.HasPrefix(line, "job_id:") {
+            parts := strings.Split(line, ":")
+            if len(parts) == 2 {
+                jobId = strings.TrimSpace(parts[1])
+                printf("JobId created for testflinger job: %s", jobId)
+                break
+            } else {
+                return nil, fmt.Errorf("JobId not found")
+            }
+        }
+    }	
+
+	s := &tfServer{
+		p: p,
+		d: tfServerData{
+			Name: name,
+			JobId: jobId,
+		},
+		system: system,
+	}
+
+	// Second step is to wait until the ip device reserved
+	printf("Waiting for testflinger %s to have an address...", name)    
+
+	timeout := time.After(30 * time.Minute)
+	warn := time.NewTicker(1 * time.Minute)
+	retry := time.NewTicker(15 * time.Second)
+	defer retry.Stop()
+    for {
+    	cmd := exec.Command("/snap/bin/testflinger-cli", "status", jobId)
+    	out, err := cmd.Output()
+    	if err != nil {
+	        return nil, fmt.Errorf("Error running command: %v", err)
+    	}
+    	line := strings.Split(string(out), "\n")[0]
+    	state := strings.TrimSpace(line)
+
+        if state == "reserve" {
+            printf("System already reserved, detecting ip...")
+            break
+        }
+
+        if state == "complete" {
+            return nil, fmt.Errorf("System already in complete state")
+        }
+        
+        select {
+		case <-retry.C:
+		case <-warn.C:
+			printf("Warning, system already on %s state", state)
+		case <-timeout:
+			s.Discard(ctx)
+			return nil, fmt.Errorf("System not reserved")
+		}
+
+    }
+
+    // wait until the ip is published (poll buffers 10 seconds until it outputs all the data)
+    time.Sleep(12 * time.Second)
```

Comment:
the comment seems disconnected (it mentions 10 seconds which is probably about the ticker below, but not why we're waiting 12 seconds here)

Reply:
Well, the test flinger client is polling every 10 seconds, so waiting 12 seconds I make sure it already made it, then I iterate every 10 seconds. If we here wait for 10 seconds should be fine as well. I am gonna update this one to 10 seconds and create a variable with the testflinger poll interval to avoid harcoding polling times in the code. Does is make sense?

Reply:
my problem was the comment itself, not the code :-) yes it makes sense